pkg/monitoring/rules/recordingrules/vmi.go

```diff
@@ -41,4 +41,44 @@ var vmiRecordingRules = []operatorrules.RecordingRule{
 		MetricType: operatormetrics.GaugeType,
 		Expr:       intstr.FromString("kubevirt_vmi_memory_available_bytes-kubevirt_vmi_memory_usable_bytes"),
 	},
+	{
+		MetricsOpts: operatormetrics.MetricOpts{
+			Name: "kubevirt_vmi_memory_used_bytes",
+			Help: "Amount of `used` memory as seen by the domain.",
+		},
+		MetricType: operatormetrics.GaugeType,
+		Expr:       intstr.FromString("kubevirt_vmi_memory_available_bytes-kubevirt_vmi_memory_usable_bytes"),
+	},
+	{
+		MetricsOpts: operatormetrics.MetricOpts{
+			Name: "kubevirt_vmi_vcpu_count",
+			Help: "The number of the VMI vCPUs.",
+		},
+		MetricType: operatormetrics.GaugeType,
+		Expr:       intstr.FromString("count by (namespace, name, node) (kubevirt_vmi_vcpu_seconds_total)"),
+	},
+	{
+		MetricsOpts: operatormetrics.MetricOpts{
+			Name: "kubevirt_vmi_guest_vcpu_queue",
+			Help: "Guest queue length.",
+		},
+		MetricType: operatormetrics.GaugeType,
+		Expr:       intstr.FromString("clamp_min(kubevirt_vmi_guest_load_1m - kubevirt_vmi_vcpu_count, 0)"),
+	},
+	{
+		MetricsOpts: operatormetrics.MetricOpts{
+			Name: "kubevirt_vmi_cpu_budget_hertz",
+			Help: "CPU budget in Hertz.",
+		},
+		MetricType: operatormetrics.GaugeType,
+		Expr:       intstr.FromString("kubevirt_vmi_vcpu_count * on(node) group_left() ( avg by(node)(node_cpu_frequency_hertz) )"),
```

Comment:
Isn't this a super misleading metric, because it ignores things like the actual CPU time that a process (i.e. qemu) gets? How does it relate to frequency scaling on a node? It can also be misleading, as the IPC of a particular CPU can vary and a system with a high IPC and lower clock frequency can still perform better than a system with a low IPC and higher clock frequency.

See e.g. https://github.com/prometheus/node_exporter/issues/2312

Reply:
I still think this is a super misleading metric. Why not simply monitor the host CPU load as percentage? A vCPU rarely maps 1:1 to the clock speed of a host CPU.